airbyte-integrations/connectors/source-google-search-console/source_google_search_console/streams.py

```diff
@@ -0,0 +1,303 @@
+#
+# MIT License
+#
+# Copyright (c) 2020 Airbyte
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+
+import json
+from abc import ABC
+from typing import Any, Dict, Iterable, List, Mapping, MutableMapping, Optional, Union
+from urllib.parse import quote_plus, unquote_plus
+
+import pendulum
+import requests
+from airbyte_cdk.sources.streams.http import HttpStream
+from airbyte_cdk.sources.streams.http.auth import HttpAuthenticator
+from airbyte_cdk.sources.streams.http.auth.oauth import Oauth2Authenticator
+from source_google_search_console.service_account_authenticator import ServiceAccountAuthenticator
+
+BASE_URL = "https://www.googleapis.com/webmasters/v3"
+GOOGLE_TOKEN_URI = "https://oauth2.googleapis.com/token"
+ROW_LIMIT = 25000
+
+
+class GoogleSearchConsole(HttpStream, ABC):
+    url_base = BASE_URL
+    primary_key = None
+    data_field = None
+
+    def __init__(
+        self,
+        client_id: str,
+        client_secret: str,
+        refresh_token: str,
+        site_urls: str,
+        start_date: pendulum.datetime,
+        service_account_info: str,
+    ):
+        super().__init__()
+        self._client_id = client_id
+        self._client_secret = client_secret
+        self._refresh_token = refresh_token
+        self._site_urls = self.get_urls_list(site_urls)
+        self._start_date = start_date
+        self._service_account_info = service_account_info
+
+    @staticmethod
+    def get_urls_list(site_urls: list) -> List[str]:
+        return list(map(quote_plus, site_urls))
+
+    @property
+    def path_param(self):
+        return self.name[:-1]
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        if self._site_urls:
```

Comment:
pagination is used when a single logical request is too big to serve in one response (e.g: if getAllUsers returns 100k users). This is using pagination to make logically unrelated requests. it should use stream slices instead. See [stream slices](https://docs.airbyte.io/contributing-to-airbyte/python/concepts/stream_slices)

I think you'll need to also separate the class hierarchy if you use slices due to how they impact incremental syncs